src/keyboardNavigation.ts

```diff
@@ -0,0 +1,82 @@
+}
+
+export const navigateUp = (path: string, shadowRoot: ShadowRoot): void => {
+    const currentFocus = <HTMLElement>shadowRoot.querySelector(`[data-path="${path}"]`);
+    if (!currentFocus || !currentFocus.parentNode) return;
+    const currentFocusAndSiblings = [...currentFocus.parentNode.children];
+    const previousSibling = getPreviousSibling(currentFocus, currentFocusAndSiblings);
+    if (previousSibling instanceof ShadowRoot) return;
+    const newFocus =
+        !isDeepestVisible(previousSibling) && !isAncestor(path, previousSibling.getAttribute('data-path') || '')
+            ? findDeepestVisibleChild(previousSibling)
+            : previousSibling;
+    if (newFocus) changeFocusedKey(newFocus, currentFocus);
+};
+
+export const navigateDown = (path: string, shadowRoot: ShadowRoot): void => {
+    const currentFocus = <HTMLElement>shadowRoot.querySelector(`[data-path="${path}"]`);
+    if (!currentFocus || !currentFocus.parentNode) return;
+    const currentFocusAndSiblings = [...currentFocus.parentNode.children];
+    const newFocus = findNextElementToFocus(currentFocus, currentFocusAndSiblings);
+    if (newFocus) changeFocusedKey(newFocus, currentFocus);
+};
+
+const findNextElementToFocus = (currentFocus: HTMLElement, currentFocusAndSiblings: Array<Element>) => {
+    if (!isDeepestVisible(currentFocus)) {
+        if (isEmptyObject(currentFocus)) {
+            return findNearestParentWithAdditionalLeaves(currentFocus);
+        } else {
+            return <HTMLElement>currentFocus.querySelector('[part=property]');
+        }
+    } else if (isLastChild(currentFocus, currentFocusAndSiblings)) {
+        return findNearestParentWithAdditionalLeaves(currentFocus) || currentFocus;
+    } else return getNextSibling(currentFocus, currentFocusAndSiblings);
+};
+
+export const navigateToParent = (path: string, shadowRoot: ShadowRoot): void => {
+    const currentFocus = <HTMLElement>shadowRoot.querySelector(`[data-path="${path}"]`);
+    if (!currentFocus || !currentFocus.parentNode || !currentFocus.parentNode.parentNode) return;
+    changeFocusedKey(<HTMLElement>currentFocus.parentNode.parentNode, currentFocus);
+};
+
+const changeFocusedKey = (newFocus: HTMLElement, oldFocus: HTMLElement): void => {
+    const keyToFocus = <HTMLSpanElement>newFocus.querySelector('[part=key]');
+    const keyToBlur = <HTMLSpanElement>oldFocus.querySelector('[part=key]');
+    keyToFocus.setAttribute('tabindex', '0');
+    keyToBlur.setAttribute('tabindex', '-1');
+    keyToFocus.focus();
+};
+
+const getNextSibling = (currentFocus: HTMLElement, currentFocusAndSiblings: Array<Element>): HTMLElement =>
+    <HTMLElement>currentFocusAndSiblings[currentFocusAndSiblings.indexOf(currentFocus) + 1];
+
+const getPreviousSibling = (currentFocus: HTMLElement, currentFocusAndSiblings: Array<Element>): HTMLElement =>
+    <HTMLElement>currentFocusAndSiblings[currentFocusAndSiblings.indexOf(currentFocus) - 1] ||
+    currentFocus.parentNode?.parentNode;
```

Comment:
this second branch of the `or` expression should really be in the caller; i don't think it makes sense to call this a sibling

Reply:
I think I understood what you meant by this, I pulled the second branch to `navigateUp` and renamed `previousSibling` there to `possibleNewFocus`

Changes applied